arch/risc-v/src/bl808/bl808_wdt.c

```diff
@@ -0,0 +1,385 @@
+ * Name: wdt_interrupt
+ *
+ * Description:
+ *   Watchdog interrupt handler. Clears the interrupt and
+ *   Calls the attached callback if there is one.
+ *
+ ****************************************************************************/
+
+static int __wdt_interrupt(int irq, void *context, void *arg)
+{
+  struct bl808_wdt_s *priv = (struct bl808_wdt_s *)arg;
+
+  /* Clear IRQ */
+
+  BL808_UNLOCK_WDT(priv->idx);
+  modifyreg32(BL808_WDT_ICLR(priv->idx), 0, WDT_CLEAR_IRQ);
+
+  if (priv->callback != NULL)
+    {
+      priv->callback(irq, context, arg);
+    }
+
+  return OK;
+}
+
+/****************************************************************************
+ * Name: bl808_wdt_start
+ *
+ * Description:
+ *   Reset the time to the current timeout and start the watchdog.
+ *
+ ****************************************************************************/
+
+int bl808_wdt_start(FAR struct watchdog_lowerhalf_s *lower)
+{
+  struct bl808_wdt_s *priv = (struct bl808_wdt_s *)lower;
+
+  /* Enable clock */
+
+  modifyreg32(BL808_TIMER_TCCR(priv->idx), WDT_CLKSEL_MASK,
+              WDT_CLK_SRC_1K << WDT_CLKSEL_SHIFT);
+
+  /* Clear counter */
+
+  BL808_UNLOCK_WDT(priv->idx);
+  modifyreg32(BL808_WDT_COUNT_CLEAR(priv->idx), 0, WDT_CLEAR_COUNT);
+  while (getreg32(BL808_WDT_COUNTER(priv->idx)) != 0);
+  BL808_UNLOCK_WDT(priv->idx);
+  modifyreg32(BL808_WDT_COUNT_CLEAR(priv->idx), WDT_CLEAR_COUNT, 0);
+
+  priv->started = true;
+  return OK;
+}
+
+/****************************************************************************
+ * Name: bl808_timer_stop
+ *
+ * Description:
+ *   Stop the watchdog.
+ *
+ ****************************************************************************/
+
+int bl808_wdt_stop(FAR struct watchdog_lowerhalf_s *lower)
+{
+  struct bl808_wdt_s *priv = (struct bl808_wdt_s *)lower;
+
+  /* WDTs are stopped by setting the input clock to NONE.
+   * This is done to allow calling watchdog_stop and then get
+   * the time left to timeout afterwards. If we used the
+   * watchdog enable bits, the counter would reset to 0 when stopped.
+   */
+
+  modifyreg32(BL808_TIMER_TCCR(priv->idx), WDT_CLKSEL_MASK,
+              WDT_CLK_SRC_NONE << WDT_CLKSEL_SHIFT);
+
+  priv->started = false;
+  return OK;
+}
+
+int bl808_wdt_keepalive(FAR struct watchdog_lowerhalf_s *lower)
+{
+  struct bl808_wdt_s *priv = (struct bl808_wdt_s *)lower;
+
+  /* Check that the watchdog is running */
+
+  if (priv->started == false)
+    {
+      return -EIO;
+    }
+
+  BL808_UNLOCK_WDT(priv->idx);
+  modifyreg32(BL808_WDT_COUNT_CLEAR(priv->idx), 0, WDT_CLEAR_COUNT);
+  while (getreg32(BL808_WDT_COUNTER(priv->idx)) != 0);
+  BL808_UNLOCK_WDT(priv->idx);
+  modifyreg32(BL808_WDT_COUNT_CLEAR(priv->idx), WDT_CLEAR_COUNT, 0);
+
+  return OK;
+}
+
+/****************************************************************************
+ * Name: bl808_wdt_getstatus
+ *
+ * Description:
+ *   Get current watchdog status. Returns to status parameter.
```

Comment:
Please include the inputs and the expected return.

Actually it always returning OK, please fix

Reply:
I'll add the inputs and return descriptions for these.

For getstatus and settimeout, these functions don't really have an error condition, so I don't see a situation where they would return an error. The prototypes in nuttx/watchdog.h are defined to return an error number, so I made these return OK to match those, which is also done by other watchdog drivers (eg bl602 or esp32c3-legacy). Is it necessary to change this?